### ISSUE TITLE
server crashing

### ISSUE THREAD

**Issue by jaxjaxk** [Zet0rz/nzombies#244]
now that we have so many main issues out of the way i believe it is time to address the elephant in the room. by elephant i mean the "ShouldCollide" hook.

we all know in the breakable entity code this hook is used. we also all know that this hook is like the dark arts of coding in gmod. this hook crashes servers frequently, you would be lucky to play the game for several hours without a crash and when a server crashes 10 times a day you know you have a real problem.to be blunt, we can NOT continue using this. so let's go over the options.

1) remove the entity all together.

2) modify the code so that the entity is SOLID_VPHYSICS when it has boards and SOLID_NONE when it does not have boards. this let's players walk through when their are no boards. to fix this we could allow players to shoot off boards so they do not get stuck. ( i know this allows for point farming but they can already do that right now)

3) find a better way. i am not the best when it comes to understanding how the world is generated but i do know that in hammer you can make a invisible wall only zombies can get through. i think it had something to do with the texture of the entity.

4) repulsion code. we could instead make it so that players that come in contact with the inadvisable wall have a physics force applied to them to make them get pushed away. so basically instead of making the object solid the players just can not force their way in. ( note: this is like an addition to 1)

bottom line, this shit crashes servers so much that i have to remove all of the barricades from all the maps.

**Comment by minin43** [comment]
My server crashes frequently for no reason I can discern. It may be this but idk

**Comment by Zet0rz** [comment]
ShouldCollide is a very dangerous hook in that it can break the physics engine, but it isn't known to crash the game (at least not afaik). Besides, we can't crash the physics engine as we don't even return anything in that hook - and it only breaks physics if you return two different values between the same two entities.

Let me adress your proposals:
1) There is **no way** we will be removing barricades from the gamemode. Removing it when it is broken in-game also removes the repair mechanic, a main mechanic of Zombies.

2) This is what it already does, except it sets itself to SOLID_VPHYSICS even when there are no planks, as long as there is no zombie going towards it. Making it permanently SOLID_NONE would allow players to go to areas they shouldn't. Invisible walls would just block the zombies in.

3) As far as I know, CLIP-brushes can only be made in Hammer as it actually compiles the clips differently during compilation; it's not just a texture that the engine handles, it's the compiler. I am really looking for a way to block out players but allow zombies, but the only way _**is** the ShouldCollide hook._ Would be awesome if they could make it in-engine though, and I have already suggested that to the gmod devs.

4) Repulsion code wouldn't work properly in such a thin entity as this. You could just force your way through. Strong repulsion would basically make it just as solid as ... well being solid, only much more laggy.

Bottom line:
You draw out so many valid points, but you just seem so misinformed. I have a lot of experience with ShouldCollide, and this isn't in any risk of the dangerous consequences from ShouldCollide since it doesn't actually return values. It might cause some crashes, so I do agree though that ShouldCollide mixed with timers is a really bad way to do it. I can probably find a way to do it with CurTime()-based Think-checks instead, but it'd still use ShouldCollide to trigger it.

**Comment by jaxjaxk** [comment]
1) i would rather not have a repair mechanic than a gamemode that crashes ever 30 min on a map that has like 10 of these.

2) i KNOW that players can get through. it is a solution, not a fix. thing is it is still better than crashing

3) sad to hear. we should all do some more investigation into it.

4) lag > crash. if you can make repulsion that is almost the same as an invisible wall then IDGAF.

bottom line, the mechanic either needs  to be fixed or removed because it is NOT acceptable to release a gamemode that crashes a server every 2 hours on a good day.

**Comment by Zet0rz** [comment]
I am all for fixing it, I think the last suggestion I had might help a bit. I don't want to resort to any of these workarounds, but an actual fix would be perfect. That being said, I'm actually quite sure this isn't your cause for crashes. It might be part of it, but I don't believe the current hook causes crashes (then it'd get reported soooo many more times).

You can in creative mode remove all barricades. See if it helps you a bit, let me know the results after extended running time. And we won't be changing anything if it means it causes misbehaving mechanics such as letting players through unless it is absolutely critical (which I don't think this is for the above reason).

> because it is NOT acceptable to release a gamemode that crashes a server every 2 hours on a good day

This is the very reason we haven't released it yet - we know it is buggy, it was warned about when you downloaded it :P

**Comment by jaxjaxk** [comment]
@Zet0rz i can not prove that the hook is the cause of the crashes however the server crashes much more frequently on maps that have the barricades than those that have none. that is why i removed the barricades a while back to check if they were the cause. when i put them back the crashing came back.

further more i used to have the invis walls use the same hook basically at some point and had even more crashes. their is a large positive correlation.

**Comment by lolleko** [comment]
How many entities do you have on those maps with barricades walls and 25 zombies? `print(#ents.GetAll())`

**Comment by jaxjaxk** [comment]
i have an idea. we should use flags for collisions

https://wiki.garrysmod.com/page/Entity/AddEFlags
http://pastebin.com/raw/V6M9MSRB

**Comment by Zet0rz** [comment]
It looks pretty fancy and could probably see some real use somewhere, but I don't think we really have the need for such a system. The two reasons being that one, this actually returns something in ShouldCollide, making it actually able to break the physics engine (or crash), and two, Nextbots (like our zombies) completely ignore ShouldCollide. They won't collide with the prop, but their pathing thinks it does, causing to just walk on the spot in front of props that are otherwise returning false on the zombies.

Or did you mean this for another appliance than the ShouldCollide replacement?

**Comment by jaxjaxk** [comment]
@Zet0rz ya so i fixed the problem and it was easier than anyone thought. it is so easy ill just explain it here.i will make a pull request later today but right now i need a haircut!

ok so here is what you do. set the collision group for

zombies:COLLISION_GROUP_INTERACTIVE_DEBRIS (nz_zombiebase.lua)[bottom of initialize and replaced in think]

humans:COLLISION_GROUP_PLAYER(sh_player_ingame.lua)[start of spawn function]

barricades:COLLISION_GROUP_INTERACTIVE_DEBRIS(enties/enties/breakable_entry/shared.lua)[in initialize function]

also note you MUST remove self:SetSolid() from the baracade initialize function as well.

for the inviswalls we use to block off zombie spawns we can do the exact same collision group and still only zombies can get through.

i am disappointed that no one tried this....

**Comment by jaxjaxk** [comment]
NOTICE: the zombies CAN walk through the barricade HOWEVER the way they are coded they will not do so until all boards are broken. i have not seen one get through until all are broken yet.

**Comment by Zet0rz** [comment]
I'm gonna make some tests with this, I think using DEBRIS on zombies would make them not collide with each other, but maybe just using INTERACTIVE will work. Don't worry about the pull request, I'll push my changes if they work :P

**Comment by Zet0rz** [comment]
I'm gonna hold back on this for now - it has an unwanted side effect of making zombies stack, making training them waaaayy too easy.
![](http://i.imgur.com/vS6Js8E.png)
(~8 zombies in that stack)

It seems like COLLISION_GROUP_INTERACTIVE doesn't collide with itself, even though the wiki would say it should - if this is a game bug, we can use this solution when that is fixed. For now I'll instead try removing the timers from ShouldCollide and see if that should make it better.

**Comment by Zet0rz** [comment]
Commit 41bf9b5b313cbf6e62c7ecb63eabf1dc66af6e87 changes the timers to CurTime() checks and removes all returning in the function. See if it causes less crashes (it should, it's now just a less called Think function).

**Comment by Zet0rz** [comment]
No words, no reports. Will consider fixed until someone reports it again.

**Comment by jaxjaxk** [comment]
@Zet0rz 
i did it my way and never had a problem again.

i highly recommend switching over to my method of using collision groups as it has never caused problems and no one has ever broken thorough the walls no matter how hard they tried. on top of that it eliminates using the should collide hook so it is less likely to fail. the zombies naturally will not stack up, the only time they do is instances like this when they have to break down a barrier. other than that the collision system pushes them away from each other naturally so it is not noticeable at all.
